tests/mr/host/mr_tests.cpp

```diff
@@ -0,0 +1,239 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+
+#include <rmm/mr/host/host_memory_resource.hpp>
+#include <rmm/mr/host/new_delete_resource.hpp>
+#include <rmm/mr/host/pinned_memory_resource.hpp>
+
+#include <cuda_runtime_api.h>
+#include <cstddef>
+#include <deque>
+#include <random>
+
+namespace {
+inline bool is_aligned(void* p,
+                       std::size_t alignment = alignof(std::max_align_t)) {
+  return (0 == reinterpret_cast<uintptr_t>(p) % alignment);
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Returns if a pointer points to a device memory or managed memory
+ * allocation.
+ *---------------------------------------------------------------------------**/
+inline bool is_device_memory(void* p) {
+  cudaPointerAttributes attributes{};
+  if (cudaSuccess != cudaPointerGetAttributes(&attributes, p)) {
+    return false;
+  }
+#if CUDART_VERSION < 10000  // memoryType is deprecated in CUDA 10
+  return attributes.memoryType == cudaMemoryTypeDevice;
+#else
+  return (attributes.type == cudaMemoryTypeDevice) or
+         (attributes.type == cudaMemoryTypeManaged);
+#endif
+}
+
+/**
+ * @brief Returns if a pointer `p` points to pinned host memory.
+ */
+inline bool is_pinned_memory(void* p) {
+  cudaPointerAttributes attributes{};
+  if (cudaSuccess != cudaPointerGetAttributes(&attributes, p)) {
+    return false;
+  }
+  return attributes.type == cudaMemoryTypeHost;
+}
+
+static constexpr std::size_t size_word{4};
+static constexpr std::size_t size_kb{std::size_t{1} << 10};
+static constexpr std::size_t size_mb{std::size_t{1} << 20};
+static constexpr std::size_t size_gb{std::size_t{1} << 30};
+static constexpr std::size_t size_tb{std::size_t{1} << 40};
+static constexpr std::size_t size_pb{std::size_t{1} << 50};
+
+struct allocation {
+  void* p{nullptr};
+  std::size_t size{0};
+  allocation(void* _p, std::size_t _size) : p{_p}, size{_size} {}
+  allocation() = default;
+};
+}  // namespace
+
+template <typename MemoryResourceType>
+struct MRTest : public ::testing::Test {
+  std::unique_ptr<rmm::mr::host_memory_resource> mr;
+
+  MRTest() : mr{new MemoryResourceType} {}
+  ~MRTest() = default;
+};
+
+using resources = ::testing::Types<rmm::mr::new_delete_resource,
+                                   rmm::mr::pinned_memory_resource>;
+
+TYPED_TEST_CASE(MRTest, resources);
+
+TYPED_TEST(MRTest, SelfEquality) { EXPECT_TRUE(this->mr->is_equal(*this->mr)); }
+
+TYPED_TEST(MRTest, AllocateZeroBytes) {
+  void* p{nullptr};
+  EXPECT_NO_THROW(p = this->mr->allocate(0));
+  EXPECT_EQ(nullptr, p);
+  EXPECT_NO_THROW(this->mr->deallocate(p, 0));
+}
+
+TYPED_TEST(MRTest, AllocateWord) {
```

Comment:
Looks like AllocateWord, AllocateKB, AllocateMB, and AllocateGB could be refactored to a single function that takes size as a param

Reply:
Good point. Could probably also re-use a bunch of existing device_memory_resource tests by parameterizing on memory_resource type?

Reply:
`device_memory_resource` and `host_memory_resource` can't really shared the same tests because `host_memory_resource` allocation uses an `alignment` parameter whereas `device_memory_resource` alignment is fixed to 256. Likewise, the device memory resource tests test for if the memory is device memory, which will obviously fail for host memory resources.

Reply:
> Looks like AllocateWord, AllocateKB, AllocateMB, and AllocateGB could be refactored to a single function that takes size as a param

I have done this in my current new suballocator branch for the device mr_tests.cpp

Reply:
> `device_memory_resource` and `host_memory_resource` can't really shared the same tests because `host_memory_resource` allocation uses an `alignment` parameter whereas `device_memory_resource` alignment is fixed to 256. Likewise, the device memory resource tests test for if the memory is device memory, which will obviously fail for host memory resources.

Both things that could be factored out / specialized.